yt/yt/server/http_proxy/component_discovery.cpp

```diff
@@ -0,0 +1,408 @@
+            simplePlural.pop_back();
+            return simplePlural;
+        }
+        case EYtComponentType::JobProxies:
+            return "job_proxy";
+        case EYtComponentType::HttpProxies:
+            return "http_proxy";
+        case EYtComponentType::RpcProxies:
+            return "rpc_proxy";
+        default:
+            YT_ABORT();
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+void Serialize(const TYtInstance& instance, IYsonConsumer* consumer)
+{
+    BuildYsonFluently(consumer)
+        .BeginMap()
+            .Item("address").Value(instance.Address)
+            .Item("type").Value(instance.Type)
+            .DoIf(instance.Error.IsOK(), [&] (auto fluent) {
+                fluent
+                    .Item("version").Value(instance.Version)
+                    .Item("start_time").Value(instance.StartTime)
+                    .Item("banned").Value(instance.Banned);
+
+                if (!instance.State.empty()) {
+                    fluent.Item("state").Value(instance.State);
+                }
+            })
+            .DoIf(!instance.Error.IsOK(), [&] (auto fluent) {
+                fluent.Item("error").Value(instance.Error);
+            })
+        .EndMap();
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+TComponentDiscoverer::TComponentDiscoverer(IClientPtr client, TMasterReadOptions masterReadOptions, TDuration proxyDeathAge)
+    : Client_(std::move(client))
+    , MasterReadOptions_(std::move(masterReadOptions))
+    , ProxyDeathAge_(proxyDeathAge)
+{ }
+
+std::vector<TYtInstance> TComponentDiscoverer::ListNodes(EYtComponentType component) const
+{
+    TListNodeOptions options;
+    FillMasterReadOptions(options, MasterReadOptions_);
+    options.Attributes = {
+        "register_time",
+        "version",
+        "banned",
+        "state",
+        "job_proxy_build_version",
+    };
+
+    auto rsp = WaitFor(Client_->ListNode(GetCypressDirectory(component), options))
+        .ValueOrThrow();
+    auto rspList = ConvertToNode(rsp)->AsList();
+
+    std::vector<TYtInstance> instances;
+    for (const auto& node : rspList->GetChildren()) {
+        auto version = node->Attributes().Find<TString>("version");
+        auto banned = node->Attributes().Find<bool>("banned");
+        auto nodeState = node->Attributes().Find<TString>("state");
+        auto startTime = node->Attributes().Find<TString>("register_time");
+
+        TYtInstance instance;
+        instance.Type = GetInstanceType(component);
+        instance.Address = node->GetValue<TString>();
+
+        if (nodeState) {
+            instance.State = *nodeState;
+        }
+        instance.Online = (nodeState == TString("online"));
+
+        if (version) {
+            instance.Version = *version;
+        }
+        if (startTime) {
+            instance.StartTime = *startTime;
+        }
+        instance.Banned = banned ? *banned : false;
```

Comment:
nip: can default value be passed to Find (or another method) above?

Reply:
There is a Get with default! Will use it as well.